[WIP] Inital pass at refactoring transaction database tests

kulpreet opened the pull request:
@evoskuil, if you are OK with these changes, I want to next write tests for other methods like candidate, uncandidate, confirm etc.

Creating this PR to get some early feedback to see if I am on the right track.

evoskuil commented:
Thanks @kulpreet ! My quick feedback is that by using the test fixture in this manner you are linking the behavior of each test. This means that one test can break other tests, and also precludes concurrent test execution. Performance and source optimization of tests are secondary goals in relation to test independence.

kulpreet commented:
My hope was that DRYing out the setup/teardown code will help write diverse test cases, but I see your point.

I'll revert the test structure and focus on test cases instead. :)

kulpreet commented:
Hi @evoskuil 

Updated this PR to revert back to keeping tests independent. Also added tests for transaction_database's store, candidate and uncandidate methods.

Let me know if I am on the right track, will keep updating this PR to include tests for the other methods in transaction_database.

evoskuil commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+    BOOST_REQUIRE(decode_base16(wire_tx2, TRANSACTION2));
+    BOOST_REQUIRE(tx2.from_data(wire_tx2));
+
+    const auto path = DIRECTORY "/tx_table";
```
I've move this into a file-scoped declaration.

kulpreet replied:
Moved, changed var name to file_path to avoid conflicting with boost::filesystem::path. There probably was a better way to resolve it, but I took the easy path - pun intended ;)

evoskuil commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+// Candidate/Uncandidate
+// ----------------------------------------------------------------------------
+
+BOOST_AUTO_TEST_CASE(transaction_database__candidate__with_no_input_in_db__returns_false)
```
Avoid acronyms (db), don't need to use the term "returns" for the result, just use "false". Really we shouldn't be using db in the tests either. We make a few exceptions for common use terms (like tx). In tests we often just use "instance" for an instance of the object under test.

kulpreet replied:
Replaced db with instance in test code.

evoskuil commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+
+}
+
+BOOST_AUTO_TEST_CASE(transaction_database__candidate__with_input_in_db__returns_true)
```
This isn't a return value, it's the value of `candidate_spend` being tested, so name:
```cpp
BOOST_AUTO_TEST_CASE(transaction_database__candidate__with_input_in_db__candidate_spend_true)
```

kulpreet replied:
Fixed, in other tests too.

evoskuil commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+    
+    chain::output::list tx1_outputs;
+    tx1_outputs.emplace_back();
+    tx1_outputs.back().set_value(1200);
```
Avoid this construct, use initializer list instead (as with `tx1_inputs` above, or for the entire `tx1` (below)).

evoskuil replied:
The reason we want to avoid updating writable members is that it breaks the caching of derived property values. The exposure of writable properties is effectively deprecated.

kulpreet replied:
Switched to initializers everywhere I create inputs and outputs.

evoskuil commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+
+    // tx2: spends coinbase, tx1 as input, dummy output
+    chain::input::list tx2_inputs;
+    chain::output_point previous_output{hash1, 0};
```
Whatespace: `previous_output{ hash1, 0 };`.

kulpreet replied:
Fixed.

evoskuil commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+
+    chain::output::list tx2_outputs;
+    tx2_outputs.emplace_back();
+    tx2_outputs.back().set_value(1200);
```
As above, avoid this form of construction, prefer the initializer.

kulpreet replied:
Done.

kulpreet commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+    BOOST_REQUIRE(result1);
+    BOOST_REQUIRE(result1.transaction().hash() == hash1);
+
+    db.store({tx2}, 200, 220);
```
I went with the list as there is no method in the class that takes a single confirmed transaction with height/fork and MTP.

Should we add such a method?

Here's the three we have for now

```
    /// Store a transaction not associated with a block.
    bool store(const chain::transaction& tx, uint32_t forks);

    /// Store a set of transactions associated with an unconfirmed block.
    bool store(const chain::transaction::list& transactions);

    /// Store a set of transactions associated with a confirmed block.
    bool store(const chain::transaction::list& transactions, size_t height,
        uint32_t median_time_past);
```

evoskuil replied:
No, since there is no scenario where we would have a single confirmation independent of the set of others in a block. So you could just just change the test to:

```cpp
db.store(tx1, 100);
...
db.store(tx2, 200)
```

evoskuil replied:
Actually there is an envisionable scenario where would would pass individual confirmed transactions independently - specifically where we are fanning out to write them concurrently. However this is not a current optimization since we are writing blocks in parallel during catch-up and we are not writing a block of confirmed transactions at all when confirming a self-mined block or compact block. This made me realize that this method is actually dead code:
```cpp
bool store(const chain::transaction::list& transactions, size_t height,
        uint32_t median_time_past);
```
So I have removed it in a pending pull request.

kulpreet replied:
Got it. Making the appropriate changes.

kulpreet replied:
@evoskuil - just thinking about what you said there about removing the method. Once removed, there is no way to "store a confirmed transaction". We will need to store a list of transactions and and then call confirm on the list.

I am facing this situation while trying to wrap up a test case for unconfirm. With the above mentioned method removed, I have no way left to store a confirmed transaction. I have to store it first, then confirm it as part of the setup.  Only then can I call unconfirm to test that unconfirm does the necessary things.

This would result in two mutexes being taken, first the root_mutex of the hash table and then the transaction database's metadata_mutex. I don't know yet if it is good or bad, just sharing the observation.

evoskuil replied:
The number of locks taken is not the important performance consideration given the overall design. The ability to download and store blocks in parallel is a greater optimization, and this implies that blocks are not yet validated. As such they cannot be confirmed when stored. Even when operating at the top of the chain it is generally the case that txs are stored unvalidated (via the tx pool or via request into that pool from a gap in a compact block) and later confirmed.

kulpreet replied:
Agree with not supporting a method only for tests. I was wondering if we might need it anyway later on, but you clarified why we won't need it.

I'll change the test to store and then confirm.

kulpreet commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+    const auto result3 = db.get(result2.link());
+    BOOST_REQUIRE(result3);
+    BOOST_REQUIRE(result3.transaction().hash() == hash2);
+
```
Fixed

kulpreet commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+
+}
+
+BOOST_AUTO_TEST_CASE(transaction_database__store__list_of_transactions__success)
```
Change made, in all other tests too.

kulpreet commented on test/databases/transaction_database.cpp:
```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+
+}
+
+BOOST_AUTO_TEST_CASE(transaction_database__store__single_unconfirmed_without_a_block__success)
```
Fixed.

kulpreet commented:
@evoskuil - let me know if there is something else you spot that needs fixing/changing.

It'll be nice to get this merged, as I want to create a new PR for tests on queries, and stop adding more to this.
